### ISSUE TITLE
Triceratops crashing when built on latest Ubuntu Studio 19.04

### ISSUE THREAD

**Issue by jhenkins** [thunderox/triceratops#5]
Hello!

I have hundreds of plugins, commercial and free. Even so, Triceratops is so unique that I love using it. Unfortunately it seems that the GUI part nowadays crashes when it is built against the dev libs on Ubuntu Studio 19.04, no matter which plugin host is used. I get the (seemingly) most meaningful output when using Jalv (output below). Please let me know if there is a better way to gather data, since I am not familiar with most dev tools used to create debugging output.

Thanks in advance for looking at this, if you happen to have some time.

Cheers!
Jan Henkins

Here is the Jalv output I have:

```
UI:           http://nickbailey.co.nr/triceratops/gui
Comm buffers: 524288 bytes
Update rate:  25.0 Hz

Plugin:       http://nickbailey.co.nr/triceratops
JACK Name:    Triceratops
Sample rate:  44100 Hz
Block length: 512 frames
MIDI buffers: 32768 bytes
JACK Name:    Triceratops
master_volume        = 0.750000
master_tune          = 0.000000
amp_drive            = 0.000000
filter_mode          = 1.000000
cutoff               = 0.500000
resonance            = 0.420004
filter_key_follow    = 0.000000
legato               = 0.000000
sync                 = 0.000000
warmth               = 1.000000
fm                   = 0.000000
panic                = 0.000000
active_one           = 1.000000
volume_one           = 1.000000
pulsewidth_one       = 0.000000
wave_one             = 0.000000
octave_one           = -2.000000
detune_one           = 0.000000
detune_centre_one    = 1.000000
inertia_one          = 0.000000
active_two           = 1.000000
volume_two           = 1.000000
pulsewidth_two       = 0.000000
wave_two             = 0.000000
octave_two           = -2.000000
detune_two           = 0.100000
detune_centre_two    = 1.000000
inertia_two          = 0.000000
active_three         = 1.000000
volume_three         = 1.000000
pulsewidth_three     = 0.000000
wave_three           = 0.000000
octave_three         = -2.000000
detune_three         = -0.100000
detune_centre_three  = 0.000000
inertia_three        = 0.000000
attack_one           = 0.000000
decay_one            = 0.750000
sustain_one          = 0.000000
release_one          = 0.300000
adsr1_route_one      = 1.000000
adsr1_route_one_dest = 0.000000
adsr1_route_two      = 0.000000
adsr1_route_two_dest = 3.000000
attack_two           = 0.000000
decay_two            = 0.500000
sustain_two          = 0.000000
release_two          = 0.000000
adsr2_route_one      = 0.400000
adsr2_route_one_dest = 1.000000
adsr2_route_two      = 0.000000
adsr2_route_two_dest = 4.000000
attack_three         = 0.000000
decay_three          = 0.000000
sustain_three        = 0.000000
release_three        = 0.000000
adsr3_lfo1_amount    = 0.000000
adsr3_lfo2_amount    = 0.000000
adsr3_lfo3_amount    = 0.000000
lfo1_retrig          = 0.000000
lfo1_speed           = 120.000000
lfo1_wave            = 0.000000
lfo1_dco1_pitch      = 0.000000
lfo1_dco2_pitch      = 0.000000
lfo1_dco3_pitch      = 0.000000
lfo1_filter          = 0.000000
lfo1_route_one       = 0.000000
lfo1_route_one_dest  = 2.000000
lfo1_route_two       = 0.000000
lfo1_route_two_dest  = 0.000000
lfo2_retrig          = 0.000000
lfo2_speed           = 120.000000
lfo2_wave            = 0.000000
lfo2_dco1_pitch      = 0.000000
lfo2_dco2_pitch      = 0.000000
lfo2_dco3_pitch      = 0.000000
lfo2_filter          = 0.000000
lfo2_route_one       = 0.000000
lfo2_route_one_dest  = 0.000000
lfo2_route_two       = 0.000000
lfo2_route_two_dest  = 0.000000
lfo3_retrig          = 0.000000
lfo3_speed           = 120.000000
lfo3_wave            = 0.000000
lfo3_dco1_pitch      = 0.000000
lfo3_dco2_pitch      = 0.000000
lfo3_dco3_pitch      = 0.000000
lfo3_filter          = 0.000000
lfo3_route_one       = 0.000000
lfo3_route_one_dest  = 0.000000
lfo3_route_two       = 0.000000
lfo3_route_two_dest  = 0.000000
fx_echo_active       = 0.000000
fx_echo_speed        = 4096.000000
fx_echo_decay        = 0.250000
fx_echo_eq_low       = 1.000000
fx_echo_eq_mid       = 1.000000
fx_echo_eq_high      = 1.000000
unison_activate      = 0.000000
unison_one           = 0.000000
unison_two           = 0.000000
unison_three         = 0.000000
modifier_dirt        = 0.000000
fx_reverb_active     = 0.000000
fx_reverb_decay      = 4.000000
fx_reverb_mix        = 0.000000
stereo_mode          = 1.000000
dco1_pan             = 0.500000
dco2_pan             = 0.500000
dco3_pan             = 0.500000
modifier_ring        = 0.000000
preset_category      = 0.000000
pitch_bend_range     = 24.000000
midi_channel         = 1.000000
lv2ui_descriptor(0) called
Loading local config file..
loading categories from /usr/local/lib/lv2/triceratops.lv2/triceratops_categories.txt
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
./start: line 1:  9743 Aborted                 (core dumped) jalv.gtk http://nickbailey.co.nr/triceratops
```

**Comment by jhenkins** [comment]
Some (hopefully useful) additional info:

Ubuntu 19.04does not have coredumps enabled by default, so after I re-learned how to get this going again with "ulimit -c unlimited", I could get jalv to create a dump. Once I had that, I could get the new-fangled systemd "coredumpctl" command to start gdb with the core and get a backtrace, which is included below.

Cheers,
Jan Henkins

gdb backtrace:

~ $ coredumpctl debug 
           PID: 15467 (jalv.gtk)
           UID: 1000 (jan)
           GID: 1000 (jan)
        Signal: 6 (ABRT)
     Timestamp: Sun 2019-06-02 11:01:29 BST (8min ago)
  Command Line: jalv.gtk http://nickbailey.co.nr/triceratops
    Executable: /usr/bin/jalv.gtk
 Control Group: /user.slice/user-1000.slice/session-c2.scope
          Unit: session-c2.scope
         Slice: user-1000.slice
       Session: c2
     Owner UID: 1000 (jan)
       Boot ID: bf8438ebf7e34c04a44c1cbe2c5d464a
    Machine ID: f57292756b214d6d9c10dae570474523
      Hostname: songbird
       Storage: /var/lib/systemd/coredump/core.jalv\x2egtk.1000.bf8438ebf7e34c04a44c1cbe2c5d464a.15467.1559469689000000.lz4
       Message: Process 15467 (jalv.gtk) of user 1000 dumped core.
                
                Stack trace of thread 15467:
                #0  0x00007f325fee5ed7 __GI_raise (libc.so.6)
                #1  0x00007f325fec7535 __GI_abort (libc.so.6)
                #2  0x00007f325fb3b672 n/a (libstdc++.so.6)
                #3  0x00007f325fb472b6 n/a (libstdc++.so.6)
                #4  0x00007f325fb47301 _ZSt9terminatev (libstdc++.so.6)
                #5  0x00007f325fb47535 __cxa_throw (libstdc++.so.6)
                #6  0x00007f325fb3dc56 _ZSt16__throw_bad_castv (libstdc++.so.6)
                #7  0x00007f325c57fd82 _ZN7presets18read_category_fileEv (triceratops_gui.so)
                #8  0x00007f325c58053e _ZN7presetsC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE (triceratops_gui.so)
                #9  0x00007f325c55fef7 _ZL8make_guiP14triceratopsGUI (triceratops_gui.so)
                #10 0x00007f325c5637f0 _ZL11instantiatePK17_LV2UI_DescriptorPKcS3_PFvPvjjjPKvES4_PS4_PKPK12_LV2_Feature (triceratops_gui.so)
                #11 0x00007f32606a0564 suil_instance_new (libsuil-0.so.0)
                #12 0x000055a2454d9d5c n/a (jalv.gtk)
                #13 0x000055a2454e10f4 n/a (jalv.gtk)
                #14 0x000055a2454d7f39 n/a (jalv.gtk)
                #15 0x00007f325fec8b6b __libc_start_main (libc.so.6)
                #16 0x000055a2454d844a n/a (jalv.gtk)
                
                Stack trace of thread 15471:
                #0  0x00007f325ffb92e9 syscall (libc.so.6)
                #1  0x00007f32606809ff n/a (libjack.so.0)
                #2  0x00007f32606687b5 n/a (libjack.so.0)
                #3  0x00007f32606678c7 n/a (libjack.so.0)
                #4  0x00007f3260667118 n/a (libjack.so.0)
                #5  0x00007f326067ec46 n/a (libjack.so.0)
                #6  0x00007f3260e33182 start_thread (libpthread.so.0)
                #7  0x00007f325ffbfb1f __clone (libc.so.6)
                
                Stack trace of thread 15468:
                #0  0x00007f3260e393bb futex_wait_cancelable (libpthread.so.0)
                #1  0x00007f326067f65c n/a (libjack.so.0)
                #2  0x00007f32606774e5 n/a (libjack.so.0)
                #3  0x00007f326067ec46 n/a (libjack.so.0)
                #4  0x00007f3260e33182 start_thread (libpthread.so.0)
                #5  0x00007f325ffbfb1f __clone (libc.so.6)
                
                Stack trace of thread 15469:
                #0  0x00007f3260e3cd94 __libc_read (libpthread.so.0)
                #1  0x00007f326067fead n/a (libjack.so.0)
                #2  0x00007f3260683301 n/a (libjack.so.0)
                #3  0x00007f326067ec46 n/a (libjack.so.0)
                #4  0x00007f3260e33182 start_thread (libpthread.so.0)
                #5  0x00007f325ffbfb1f __clone (libc.so.6)
                
                Stack trace of thread 15470:
                #0  0x00007f3260e3cd94 __libc_read (libpthread.so.0)
                #1  0x00007f326067fead n/a (libjack.so.0)
                #2  0x00007f3260683301 n/a (libjack.so.0)
                #3  0x00007f326067ec46 n/a (libjack.so.0)
                #4  0x00007f3260e33182 start_thread (libpthread.so.0)
                #5  0x00007f325ffbfb1f __clone (libc.so.6)

GNU gdb (Ubuntu 8.2.91.20190405-0ubuntu3) 8.2.91.20190405-git
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/jalv.gtk...
(No debugging symbols found in /usr/bin/jalv.gtk)
[New LWP 15467]
[New LWP 15471]
[New LWP 15468]
[New LWP 15469]
[New LWP 15470]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `jalv.gtk http://nickbailey.co.nr/triceratops'.
Program terminated with signal SIGABRT, Aborted.
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
[Current thread is 1 (Thread 0x7f325dd0fa00 (LWP 15467))]
(gdb) quit

**Comment by millerthegorilla** [comment]
I have the same problem with the code that I downloaded from sourceforge, but trying this code from github now, I get the message: "unable to find plugin", when running 
    jalv.gtk http://nickbailey.co.nr/triceratops
and I have tried linking to the /usr/lib/lv2 directory and also moving the triceratops.lv2 directory there.  I also can't see the triceratops-presets.lv2 directory which was created when using waf before.

**Comment by thunderox** [comment]
Fixed the "std::bad_cast" problem on 19.04+ hopefully.

**Comment by jhenkins** [comment]
@thunderox Thank you very much indeed! I tested it on my 19.04 install of Ubuntu Studio, and it now works! I tested it with the following hosts:
* Jalv (works)
* Ardour 5.12 (works)
* Ardour 6 pre0 (works)
* Carla 2.0 (works)
* Harrison Mixbus 5.2.191 (partially - cannot start the GUI - error: [ERROR]: failed to instantiate LV2 GUI)

All in all I am very happy, closing this ticket as it works for me.

Cheerio,
Jan Henkins
